src/dotnet/ShellShimMaker/ShellShimMaker.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Text;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.Extensions.EnvironmentAbstractions;
+
+namespace Microsoft.DotNet.ShellShimMaker
+{
+    public class ShellShimMaker
+    {
+        private readonly string _pathToPlaceShim;
+
+        public ShellShimMaker(string pathToPlaceShim)
+        {
+            _pathToPlaceShim =
+                pathToPlaceShim ?? throw new ArgumentNullException(nameof(pathToPlaceShim));
+        }
+
+        public void CreateShim(string packageExecutablePath, string shellCommandName)
+        {
+            var packageExecutable = new FilePath(packageExecutablePath);
+
+            var script = new StringBuilder();
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                script.AppendLine("@echo off");
+                script.AppendLine($"dotnet {packageExecutable.ToQuotedString()} %*");
+            }
+            else
+            {
+                script.AppendLine("#!/bin/sh");
+                script.AppendLine($"dotnet {packageExecutable.ToQuotedString()} \"$@\"");
+            }
+
+            var scriptPath = GetScriptPath(shellCommandName);
+            try
+            {
+                File.WriteAllText(scriptPath.Value, script.ToString());
+            }
+            catch (UnauthorizedAccessException e)
+            {
+                throw new GracefulException(
+                    string.Format("Install failed, try run the install command as an administrator, if you have the access. {0}",
+                        e.Message));
+            }
+
+            SetUserExecutionPermissionToShimFile(scriptPath);
+        }
+
+        public void EnsureCommandNameUniqueness(string shellCommandName)
+        {
+            if (File.Exists(Path.Combine(_pathToPlaceShim, shellCommandName)))
+            {
+                throw new GracefulException($"Failed to create tool {shellCommandName}, a command with the same name existed");
```

Comment:
`a command with the same name already exists.`.

cc @KathleenDollard

Reply:
Failed to install tool {shellCommandName}. A command with the same name exists.

Reply:
Changed to Kathleen's version